File: src/analyzers/matrix.py
from __future__ import annotations

from pathlib import Path

from src.parsers.dotenv import parse_dotenv


def matrix_dotenv_files(
    paths: tuple[str, ...] | list[str],
    *,
    show_all: bool = False,
) -> dict[str, object]:
    normalized_paths = tuple(str(Path(path)) for path in paths)
    parsed = [(path, parse_dotenv(path)) for path in normalized_paths]
    latest_by_path = {path: _latest_by_name(result.definitions) for path, result in parsed}
    definitions_by_path = {path: result.definitions for path, result in parsed}
    all_names = sorted({name for definitions in latest_by_path.values() for name in definitions})

    variables = []
    inconsistent_count = 0
    for name in all_names:
        files = []
        kinds = set()
        present_in = []
        missing_in = []
        duplicates_in = []

        for path in normalized_paths:
            definition = latest_by_path[path].get(name)
            if definition is None:
                files.append({"path": path, "presence": "missing"})
                missing_in.append(path)
                continue

            kinds.add(definition.normalized_value_kind)
            present_in.append(path)
            if any(
                candidate.name == name and candidate.is_duplicate
                for candidate in definitions_by_path[path]
            ):
                duplicates_in.append(path)
            files.append(
                {
                    "path": path,
                    "presence": "present",
                    "value_kind": definition.normalized_value_kind,
                    "is_duplicate": path in duplicates_in,
                }
            )

        inconsistent = (
            len(missing_in) > 0 or len(kinds) > 1 or len(duplicates_in) > 0
        )
        if inconsistent:
            inconsistent_count += 1

        variable = {
            "name": name,
            "status": "inconsistent" if inconsistent else "consistent",
            "present_in": present_in,
            "missing_in": missing_in,
            "duplicates_in": duplicates_in,
            "value_kinds": sorted(kinds),
            "files": files,
        }
        if show_all or inconsistent:
            variables.append(variable)

    warnings = sorted(
        warning for _, result in parsed for warning in result.warnings
    )
    return {
        "paths": list(normalized_paths),
        "show_all": show_all,
        "file_count": len(normalized_paths),
        "variable_count": len(all_names),
        "inconsistent_variable_count": inconsistent_count,
        "variables": variables,
        "warnings": warnings,
    }


def _latest_by_name(definitions) -> dict[str, object]:
    latest = {}
    for definition in definitions:
        latest[definition.name] = definition
    return latest

File: internal/analyzers/matrix.go
package analyzers

import (
	"github.com/davisbuilds/envdiff/internal/dotenv"
	"github.com/davisbuilds/envdiff/internal/model"
	"github.com/davisbuilds/envdiff/internal/order"
)

func MatrixDotenvFiles(paths []string, showAll bool) (map[string]any, error) {
	type parsedFile struct {
		path   string
		result model.DotenvParseResult
	}

	parsed := []parsedFile{}
	latestByPath := map[string]map[string]model.EnvVarDefinition{}
	definitionsByPath := map[string][]model.EnvVarDefinition{}
	allNameSet := map[string]struct{}{}
	for _, path := range paths {
		result, err := dotenv.Parse(path)
		if err != nil {
			return nil, err
		}
		parsed = append(parsed, parsedFile{path: path, result: result})
		latest := latestByName(result.Definitions)
		latestByPath[path] = latest
		definitionsByPath[path] = result.Definitions
		for name := range latest {
			allNameSet[name] = struct{}{}
		}
	}

	allNames := make([]string, 0, len(allNameSet))
	for name := range allNameSet {
		allNames = append(allNames, name)
	}
	allNames = order.Strings(allNames)

	variables := []map[string]any{}
	inconsistentCount := 0
	for _, name := range allNames {
		files := []map[string]any{}
		kindSet := map[string]struct{}{}
		presentIn := []string{}
		missingIn := []string{}
		duplicatesIn := []string{}

		for _, path := range paths {
			definition, ok := latestByPath[path][name]
			if !ok {
				files = append(files, map[string]any{"path": path, "presence": "missing"})
				missingIn = append(missingIn, path)
				continue
			}

			kindSet[definition.NormalizedValueKind] = struct{}{}
			presentIn = append(presentIn, path)
			isDuplicate := hasDuplicateDefinition(definitionsByPath[path], name)
			if isDuplicate {
				duplicatesIn = append(duplicatesIn, path)
			}
			files = append(files, map[string]any{
				"is_duplicate": isDuplicate,
				"path":         path,
				"presence":     "present",
				"value_kind":   definition.NormalizedValueKind,
			})
		}

		valueKinds := make([]string, 0, len(kindSet))
		for kind := range kindSet {
			valueKinds = append(valueKinds, kind)
		}
		valueKinds = order.Strings(valueKinds)
		inconsistent := len(missingIn) > 0 || len(valueKinds) > 1 || len(duplicatesIn) > 0
		if inconsistent {
			inconsistentCount++
		}

		variable := map[string]any{
			"duplicates_in": duplicatesIn,
			"files":         files,
			"missing_in":    missingIn,
			"name":          name,
			"present_in":    presentIn,
			"status":        "consistent",
			"value_kinds":   valueKinds,
		}
		if inconsistent {
			variable["status"] = "inconsistent"
		}
		if showAll || inconsistent {
			variables = append(variables, variable)
		}
	}

	warnings := []string{}
	for _, item := range parsed {
		warnings = append(warnings, item.result.Warnings...)
	}

	return map[string]any{
		"file_count":                  len(paths),
		"inconsistent_variable_count": inconsistentCount,
		"paths":                       append([]string(nil), paths...),
		"show_all":                    showAll,
		"variable_count":              len(allNames),
		"variables":                   variables,
		"warnings":                    order.Strings(warnings),
	}, nil
}

func hasDuplicateDefinition(definitions []model.EnvVarDefinition, name string) bool {
	for _, definition := range definitions {
		if definition.Name == name && definition.IsDuplicate {
			return true
		}
	}
	return false
}
